src/lib/Parser.hs

```diff
@@ -336,20 +336,23 @@ decl = do
 interfaceInstance :: Parser UDecl
 interfaceInstance = do
   keyWord InstanceKW
-  (lhs, pos) <- withPos simpleLet
+  (p, pos) <- withPos $ letPat <|> parens pat
+  ann <- annot uType
   keyWord WhereKW
-  rhs <- withSrc $ URecord <$> interfaceRecordFields "="
-  let ULet _ (_, Just (WithSrc _  ann)) _ = lhs rhs
-  let ULet _ (p, ann') rhs' = lhs $ mkRhs ann rhs
-  let (ann'', rhs'') = addImplicitImplicitArgs pos ann' rhs'
-  return $ ULet InstanceLet (p, ann'') rhs''
+  rhs <- mkRhs ann <$> (withSrc $ URecord <$> interfaceRecordFields "=")
+  let (ann', rhs') = addImplicitImplicitArgs pos (Just ann) rhs
+  return $ ULet InstanceLet (p, ann') rhs'
   where
-    mkRhs (UPi _ _ (WithSrc _ lrhs)) rhs = mkRhs lrhs rhs
-    mkRhs (UApp _ (WithSrc _ (UVar v)) _) rhs =
-      (var . nameToStr . mkInterfaceConsName . varName) v `mkApp` rhs
-    -- We let this case through to fail during type checking.
-    mkRhs _ rhs = rhs
-    var s = WithSrc Nothing $ UVar $ mkName s :> ()
+    mkRhs (WithSrc _ ann) = mkRhs' ann
+      where
+        -- Here, we are traversing the type annotation to retrieve the name of
+        -- the interface and generate its corresponding constructor.
+        mkRhs' (UPi _ _ (WithSrc _ lrhs))      rhs = mkRhs' lrhs rhs
+        mkRhs' (UApp _ (WithSrc _ (UVar v)) _) rhs =
+          (var . nameToStr . mkInterfaceConsName . varName) v `mkApp` rhs
+        -- We let this case through to fail during type checking.
+        mkRhs' _ rhs = rhs
```

Comment:
Wouldn't it produce a super confusing type error? It would be better to provide a good error message here

Reply:
My understanding is that this would only happen in cases like:
```hs
interface InterfaceTest3 _:Type where
  att : Int32

instance instanceTest3 : InterfaceTest3 where
  att = \x. 1
```
where you end up having no parameterized type. Then, this is the error you get:
```
> Type error:
> Expected: Type
>   Actual: (Type -> Type)
>
> instance instanceTest3 : InterfaceTest3 where
> 
```
I thought that this was clear enough to defer to type checking?

Reply:
Ok that makes sense I think

Reply:
Though the `Expected` and `Actual` are swapped I think?

Reply:
Hmm no, I think it is fine?
We expect to have something like `InterfaceTest3 Int32` which has type `Type`, but `InterfaceTest3` actually has type `Type -> Type`?

Reply:
Oh, right. Fair enough!